File: Tank_Game/Assets/Universal Assets/ZDelete/Missile.cs
using UnityEngine;
using System.Collections;

[RequireComponent(typeof(PlayerBasic))]

public class Missile : MonoBehaviour {
	
	public bool homing = true;
	public float damage = 1;
	public float initialSpeed = 1;
	public float fuel = Mathf.Infinity;		// How many seconds can we move before we stop?
	public float fuse = Mathf.Infinity;		// How many seconds before we blow up?
	
	private Transform targetingInner;
	private Transform targetingOuter;
	private PlayerBasic basic;

	private void Start () {
		homing = true;
		(targetingInner = (Instantiate(AssetLoad.assetLoad.targetingInner, transform.position, transform.rotation) as Transform)).parent = transform;
		(targetingOuter = (Instantiate(AssetLoad.assetLoad.targetingOuter, transform.position, transform.rotation) as Transform)).parent = transform;
		(basic = GetComponent(typeof(PlayerBasic)) as PlayerBasic).autopilot = true;
		float ratio = (float)Screen.width/(float)Screen.height;
		Camera missileCamera = (Instantiate(AssetLoad.assetLoad.newGameObject, transform.position, transform.rotation) as GameObject).AddComponent(typeof(Camera)) as Camera;
		missileCamera.transform.parent = transform;
		missileCamera.orthographicSize = 128;
		missileCamera.rect = new Rect(1-(0.4F/ratio), 0.6F, 0.4F/ratio, 0.4F);		// new Rect(0.8, 0.8, 0.2 0.2)?
		missileCamera.name = "Camera";
	}
	private void Update () {
		if ((fuse -= Time.deltaTime)<=0) {
			Explode(transform.position, null);
		}
		Vector3 rotate = Vector3.zero;
//		Vector3 translate = Vector3.zero;
		if (Mathf.Abs(basic.moveSpeed)==Mathf.Infinity) {
			RaycastHit hit;
			if (Physics.Raycast(transform.position, transform.forward, out hit, Mathf.Infinity)) {
				Debug.DrawLine(transform.position, hit.point, Color.magenta);
				PlayerBasic found = hit.transform.GetComponent(typeof(PlayerBasic)) as PlayerBasic;
				if (found) {
					DoDamage(transform.position);
					//found.SendMessage("ChangeHealth", bulletDamage);			// Tells it to update the TextMesh.
				}
			}
		}
		if (basic.autopilot) {
			Vector3 destination = (homing ? basic.GetDestination() : transform.position+transform.forward);
			bool visible = (destination!=Vector3.zero);
			targetingInner.position = destination;
			targetingOuter.position = destination;
			targetingInner.Rotate(0, 0, -180*Time.deltaTime);
			targetingOuter.Rotate(0, 0, 180*Time.deltaTime);
			targetingInner.GetComponent<Renderer>().enabled = visible;
			targetingOuter.GetComponent<Renderer>().enabled = visible;
			rotate = basic.RelativeRotationToWorldPoint((destination!=Vector3.zero ? destination : Vector3.zero), Vector3.up);
//			translate = transform.InverseTransformPoint(destination!=Vector3.zero ? destination : transform.forward);
		}
		else {
			rotate = new Vector3(Input.GetAxisRaw("Pitch"), Input.GetAxisRaw("Yaw"), Input.GetAxisRaw("Roll"))*basic.turnSpeed;
//			translate = new Vector3(Input.GetAxisRaw("Slip"), Input.GetAxisRaw("Heave"), Input.GetAxisRaw("Surge"))*basic.moveSpeed;
		}
		if (fuel>0) {
			fuel -= Time.deltaTime;
			basic.InitiateCommand(rotate, Vector3.forward);
		}
	}
	private void OnCollisionEnter (Collision collision) {
		Explode(collision.contacts[0].point, collision);
	}	
	private void DoDamage (Vector3 location) {
//		other.health-damage;		// Do Damage.
		Destroy(gameObject);
	//	Instantiate(this, location, transform.rotation);
	}

	private float power = 1;
	private float damageRange = 1;
//	private float damageMax = 1;

//	private void OnCollisionEnter (Collision collision) {		// Advanced explosion that starts at missile contact point and decreases damage the farther away object is which is determined by shooting a raycast all at it.
//	// finding the object we want, then using that point to put into damage vs distance formula.		// NOTE. I don't realy know how well this works but it seams to work fairly well.
//		ContactPoint contactPoint = collision.contacts[0];
//		Vector3 contactPosition = contactPoint.point;
//		//Instantiate(explosionPrefab, contactPosition, contactRotation);
//	 	Destroy(gameObject);
//	}
	
	
	private void Explode (Vector3 explodeOrigin, Collision collision) {
		if (damageRange<=0) {		// Does not explode.
			// Get other and do damage.
		}
		else {
			//Instantiate(explosionPrefab, contactPosition, contactRotation);
//			//Instantiate(prettyParticleExplodSphere);
			Collider[] colliders = Physics.OverlapSphere(explodeOrigin, damageRange);
			for (int j=0; j<colliders.Length; j++) {
				Collider impact = colliders[j];
				if (impact.GetComponent<Rigidbody>()) {
					impact.GetComponent<Rigidbody>().AddExplosionForce(power, explodeOrigin, damageRange);	// Exposive force.
				}
				PlayerBasic found = impact.transform.GetComponent(typeof(PlayerBasic)) as PlayerBasic;
				if (found) {
//					found.SendMessage("ChangeHealth", Mathf.Clamp01(1-Vector3.Distance(explodeOrigin, impact.transform.position)/damageRange)*damageMax);		// Objects recive less damage the farther away they are.
					//(found ? found : find).SendMessage("ChangeHealth", Mathf.Lerp(0, damageMax, (damageRange-Vector3.Distance(hit.point, contactPosition)/damageRange))
					////		other.health-damage;		// Do Damage.
				}
			}
		}
		Destroy(gameObject);
	}

}

File: Tank_Game/Assets/Universal Assets/ZDelete/Missile.js
#pragma strict
#pragma implicit
#pragma downcast

var damageMax : int;
var speedTranslate : float;
var speedRotate : float;
var homing : boolean;
var damageRange = 0.5;
var power = 10.0;
var sightRange = 50;

private var team : Transform;
private var fly : ConstantForce;
var basic : PlayerBasic;		// Public so hlicopter can reach it.

var speedfly : Vector3;

function Start () {
	basic = GetComponent(PlayerBasic);
//	basic.SendMessage("SetValue", Vector2(speedRotate, speedTranslate));
//	var useInput = new boolean[6];
//	if (guiding||homing) {
//		useInput[0] = true;
//		useInput[1] = true;
//		useInput[2] = true;
//	}
//	useInput[3] = false;
//	useInput[4] = false;
//	useInput[5] = true;
//	basic.SendMessage("SetControls", useInput);
	basic = GetComponent(PlayerBasic);
	if (basic.canControl) {
		sightRange = basic.sightRange;
		team = basic.transform.parent;
		var ratio = (0.0+Screen.width)/Screen.height;
		var missileCamera : Camera = GetComponentInChildren(Camera);
		if (missileCamera) {
			missileCamera.orthographicSize = 128;
			missileCamera.rect = Rect(1-(0.4/ratio), 0.6, 0.4/ratio, 0.4);
		}
	}
	//SendMessage("CanControl", false);
	//fly = GetComponent(ConstantForce);
	//fly.relativeForce = speedfly;
	//rigidbody.velocity = (transform.TransformDirection(speedfly));
}
	
function OnCollisionEnter (other : Collision) {		// Advanced explosion that starts at missile contact point and decreases damage the farther away object is which is determined by shooting a raycast all at it,
// finding the object we want, then using that point to put into damage vs distance formula.		// NOTE. I don't realy know how well this works but it seams to work fairly well.
	var contactPoint = other.contacts[0];
	var contactRotation = Quaternion.FromToRotation(Vector3.up, contactPoint.normal);
	var contactPosition = contactPoint.point;
	var colliders : Collider[] = Physics.OverlapSphere(contactPosition, damageRange);
	var sendDamage : float;		// Objects recive less damage the farther away they are.
	for (var impact in colliders) {
		if (impact.rigidbody) {
			impact.rigidbody.AddExplosionForce(power, contactPosition, damageRange);	// Exposive force.
		}
		if (impact.transform.parent) {
			if (transform.parent!=impact.transform.parent) {	// See if what we impact is on our team.
				var found : PlayerBasic = impact.transform.GetComponent(PlayerBasic);
				if (found) {
					var hits : RaycastHit[];
					transform.LookAt(impact.transform.position);
					hits = Physics.RaycastAll (transform.position, transform.forward, damageRange);
					for (var i=0;i<hits.length;i++) {
						var hit : RaycastHit = hits[i];
						if (impact.transform==hit.transform) {
							found.SendMessage("ChangeHealth", Mathf.Clamp01((damageRange-Vector3.Distance(hit.point, contactPosition))/damageRange)*damageMax);		// Apply less damage the farther away hit point.
							//(found ? found : find).SendMessage("ChangeHealth", Mathf.Lerp(0, damageMax, (damageRange-Vector3.Distance(hit.point, contactPosition)/damageRange));
						}
					}
				}
			}
		}
	}
	//Instantiate(explosionPrefab, contactPosition, contactRotation);
 	Destroy(gameObject);
}

function Update () {
	if (!GameController.paused) {
		transform.Translate(0, 0, Time.deltaTime*speedTranslate);
//		var xRotate : float;
//		var yRotate : float;
//		var zRotate : float;
//		if (basic.canControl&&homing) {
//			var oldAngleX = transform.eulerAngles.x;		// Save current rotation.
//			var oldAngleY = transform.eulerAngles.y;		// Save current rotation.
//			var oldAngleZ = transform.eulerAngles.z;		// Save current rotation.
//			if (homing) {
//				if (!basic.target) {		// Merge with top? If no target then try to find one.
//					var colliderList = Physics.OverlapSphere(transform.position, sightRange);
//					var objectDistance = Mathf.Infinity;
//					var ourPosition = transform.position;
//					var otherDamage : int;
//					var otherHealth : int;
//					var RotateX : float;
//					var enemyRangeValue : int;		// 
//					for (i=0; i<colliderList.length; i++) {															// Finds all objects near us.
//						var other : Collider = colliderList[i];
//						if (other.CompareTag("Character")||other.CompareTag("Defense")||other.CompareTag("Building")) {		// If they can be damaged.
//							if (team!=other.transform.parent) {
//								var obj : PlayerBasic = other.GetComponent(PlayerBasic);
//								if (obj) {
//									var enemyNear : Collider[] = Physics.OverlapSphere(other.transform.position, damageRange);
//									var price : int;
//									for (near in enemyNear) {
//										if (near.transform.parent) {
//											if (transform.parent!=near.transform.parent) {	// See if what we impact is on our team.
//												price += (near.CompareTag("Objective") ? 4 : (near.CompareTag("Character") ? 3 : (near.CompareTag("Defense") ? 2 : (near.CompareTag("Building") ? 1 : 0))));
//											}
//										}
//									}
//									if (price>enemyRangeValue-(enemyRangeValue/20)) {
//										enemyRangeValue = (price>enemyRangeValue? price : enemyRangeValue);
//										if (obj.healthInt>otherHealth-(otherHealth/10)) {
//											otherHealth = (obj.healthInt>otherHealth ? obj.healthInt : otherHealth);		// Update max health thing.
//											var difference = Vector3.Distance(other.transform.position, transform.position);
//											if (difference-(sightRange/10)<objectDistance) {						// Find the characters that are closest.
//												objectDistance = (difference<objectDistance ? difference : objectDistance);		// Have the one that is the closest be the reference.
//												transform.LookAt(other.transform);
//												var hity : RaycastHit;
//												if (Physics.Raycast(transform.position, transform.TransformDirection(Vector3.forward), hity, Mathf.Infinity)) {
//													if (hity.transform==other.transform) {					// If we are looking right at our potential target.
//														var newX = transform.eulerAngles.x;
//														if (Mathf.Abs((newX-oldAngleX)-180)+10>RotateX) {				// Target the one in front of us.
//															RotateX = Mathf.Abs(oldAngleX-180);
//															basic.target = other.transform;						// If it makes it this far then assign the object to the target.
//														}
//													}
//												}
//											}
//										}
//									}
//								}
//							}
//						}
//					}
//				}
//			}
//			if (basic.canControl) {
//				if (basic.human) {
//					xRotate = oldAngleX+(Input.GetAxis("SpinX")* Time.deltaTime * speedTranslate);
//					yRotate = oldAngleX+(Input.GetAxis("SpinY")* Time.deltaTime * speedTranslate);
//					zRotate = oldAngleX+(Input.GetAxis("SpinZ")* Time.deltaTime * speedTranslate);
//				}
//				else if (!basic.target) {
//					transform.rotation = Quaternion.Euler(oldAngleX, oldAngleY, 0);
//					transform.LookAt(basic.target);
//					var newAngleX = transform.eulerAngles.x;						// Save new rotation.
//					var newAngleY = transform.eulerAngles.y;						// Save new rotation.
//					var spinX = newAngleX+(180-oldAngleX);							// Spin oldAngle to 180 and Convert newAngle to corisponding rotation.
//					var spinY = newAngleY+(180-oldAngleY);							// Spin oldAngle to 180 and Convert newAngle to corisponding rotation.
//					spinX = (spinX>360 ? spinX-360 : (spinX<0 ? spinX+360 : spinX));
//					spinY = (spinY>360 ? spinY-360 : (spinY<0 ? spinY+360 : spinY));
//					spinX = Mathf.RoundToInt(spinX);	// Round value off to see if we are looking at it;
//					spinY = Mathf.RoundToInt(spinY);	// Round value off to see if we are looking at it;
//					var inputX = (spinX>180 ? 1 : (spinX<180 ? -1 : 0));		// This tests to see which direction to turn to reach newAngle fastest.
//					var inputY = (spinY>180 ? 1 : (spinY<180 ? -1 : 0));		// This tests to see which direction to turn to reach newAngle fastest.
//					var turnX = oldAngleX+(inputX * Time.deltaTime * (Mathf.Abs(180-spinX)<10 ? (Mathf.Abs(180-spinX)*(speedTranslate/10)*speedTranslate)/speedTranslate : speedTranslate*2));			// Get oldAngle and add speedTranslate and direction to get a new angle.
//					var turnY = oldAngleY+(inputY * Time.deltaTime * (Mathf.Abs(180-spinY)<10 ? (Mathf.Abs(180-spinY)*(speedTranslate/10)*speedTranslate)/speedTranslate : speedTranslate*2));			// Get oldAngle and add speedTranslate and direction to get a new angle.
//					turnX = Mathf.Clamp(turnX, (inputX>0 ? oldAngleX : (newAngleX>oldAngleX ? newAngleX-360 : newAngleX)), (inputX>0 ? (newAngleX<oldAngleX ? newAngleX+360 : newAngleX) : oldAngleX));		// Depending on which way we turn we clamp it.
//					turnY = Mathf.Clamp(turnY, (inputY>0 ? oldAngleY : (newAngleY>oldAngleY ? newAngleY-360 : newAngleY)), (inputY>0 ? (newAngleY<oldAngleY ? newAngleY+360 : newAngleY) : oldAngleY));		// Depending on which way we turn we clamp it.
//					xRotate = turnX;
//					yRotate = turnY;
//					zRotate = 0;
//				}
//				//transform.rotation = Quaternion.Euler(Vector3(xRotate, yRotate, zRotate));
//			}
//		}
	}
}

function OnDrawGizmos () {
	if (homing) {
		Gizmos.color = Color.cyan;
		Gizmos.DrawWireSphere(transform.position, sightRange);
	}
	Gizmos.color = Color(0.2, 0.95, 0.6);
	Gizmos.DrawWireSphere(transform.position, damageRange);
}

@script RequireComponent(ConstantForce)